File: app/src/main/java/ru/art2000/calculator/currency_converter/view/CurrenciesSettingsActivity.java
package ru.art2000.calculator.currency_converter.view;

import android.animation.Animator;
import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleEventObserver;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.viewpager.widget.ViewPager;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.card.MaterialCardView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.schedulers.Schedulers;
import ru.art2000.calculator.R;
import ru.art2000.calculator.currency_converter.model.CurrencyItem;
import ru.art2000.calculator.currency_converter.view_model.CurrenciesSettingsModel;
import ru.art2000.calculator.currency_converter.view_model.CurrencyDependencies;
import ru.art2000.extensions.DayNightActivity;
import ru.art2000.helpers.AndroidHelper;
import ru.art2000.helpers.CurrencyValuesHelper;
import ru.art2000.helpers.PrefsHelper;
import ru.art2000.helpers.SnackbarThemeHelper;

public class CurrenciesSettingsActivity extends DayNightActivity {

    public CurrenciesAddFragment add = new CurrenciesAddFragment();
    public CurrenciesEditFragment edit = new CurrenciesEditFragment();
    Context mContext;
    boolean changeDone = false;
    FloatingActionButton fab;
    MenuItem deselect;
    MenuItem select;

    TabLayout tabs;

    int selectedTab = 0;
    @DrawableRes
    int checkDrawable = R.drawable.ic_currencies_done;
    @DrawableRes
    int deleteDrawable = R.drawable.ic_clear_history;
    @DrawableRes
    int currentDrawable = checkDrawable;

    boolean isFirstTimeTooltipShown = !PrefsHelper.isDeleteTooltipShown();
    Snackbar deleteTooltip;

    CoordinatorLayout coordinatorLayout;
    LinearLayout searchViewLayout;
    SearchView barSearchView;

    boolean useViewPager2 = false;
    boolean optionsMenuCreated = false;


    CurrenciesSettingsModel model;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        setTheme(PrefsHelper.getAppTheme());
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_currencies_editor);
        mContext = this;
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        model = new CurrenciesSettingsModel(this.getApplication());

        fab = findViewById(R.id.floatingActionButton);
        fab.addOnShowAnimationListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
                fab.setImageResource(currentDrawable);
            }

            @Override
            public void onAnimationEnd(Animator animator) {

            }

            @Override
            public void onAnimationCancel(Animator animator) {
                fab.setImageResource(currentDrawable);
            }

            @Override
            public void onAnimationRepeat(Animator animator) {

            }
        });
        fab.addOnHideAnimationListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {

            }

            @Override
            public void onAnimationEnd(Animator animator) {
                if (currentDrawable == deleteDrawable) {
                    showDeleteTip();
                }
            }

            @Override
            public void onAnimationCancel(Animator animator) {

            }

            @Override
            public void onAnimationRepeat(Animator animator) {

            }
        });

        coordinatorLayout = findViewById(R.id.coordinator);
        searchViewLayout = findViewById(R.id.search_view_layout);
        barSearchView = findViewById(R.id.search_view);

        tabs = findViewById(R.id.tabs);

        ViewPager pager = findViewById(R.id.pager);
        ViewPager2 pager2 = findViewById(R.id.pager2);

        if (useViewPager2) {
            pager.setVisibility(View.GONE);

            pager2.setOffscreenPageLimit(2);
            pager2.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
                @Override
                public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                    int maxScroll = pager2.getMeasuredWidth();
                    int currentScroll = maxScroll * position + positionOffsetPixels;
                    searchViewLayout.setTranslationX(-currentScroll);
                    if (deleteTooltip != null) {
                        deleteTooltip.getView().setTranslationX(maxScroll - currentScroll);
                    }
                }
            });
            CurrencyEditorPager2Adapter pager2Adapter = new CurrencyEditorPager2Adapter();
            pager2.setAdapter(pager2Adapter);
            new TabLayoutMediator(tabs, pager2, (tab, position) ->
                    tab.setText(pager2Adapter.getPageTitle(position))).attach();
        } else {

            pager2.setVisibility(View.GONE);

            pager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
                @Override
                public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                    int maxScroll = pager.getMeasuredWidth();
                    int currentScroll = maxScroll * position + positionOffsetPixels;
                    searchViewLayout.setTranslationX(-currentScroll);
                    if (deleteTooltip != null) {
                        deleteTooltip.getView().setTranslationX(maxScroll - currentScroll);
                    }
                }

                @Override
                public void onPageSelected(int position) {

                }

                @Override
                public void onPageScrollStateChanged(int state) {

                }
            });
            pager.setAdapter(new CurrencyEditorPagerAdapter(getSupportFragmentManager()));
            tabs.setupWithViewPager(pager);
        }

        tabs.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                selectedTab = tab.getPosition();
                fab.hide();
                modifyVisualElements(tab.getPosition());
                if (optionsMenuCreated) {
                    toggleElementsVisibility();
                }
                if (selectedTab == 1 && !fab.isShown()) {
                    showDeleteTip();
                } else if (selectedTab == 0 && deleteTooltip != null) {
                    deleteTooltip.dismiss();
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
                switch (tab.getPosition()) {
                    default:
                    case 0:
                        add.scrollToTop();
                        break;
                    case 1:
                        edit.scrollToTop();
                        break;
                }
            }

        });
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        searchViewLayout.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED);
        add.recyclerViewBottomPadding = searchViewLayout.getMeasuredHeight();

        CoordinatorLayout.LayoutParams layoutParams = (CoordinatorLayout.LayoutParams) fab.getLayoutParams();
        layoutParams.bottomMargin = add.recyclerViewBottomPadding;
        fab.setLayoutParams(layoutParams);

        selectedTab = tabs.getSelectedTabPosition();
        modifyVisualElements(selectedTab);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.currencies_editor_menu, menu);

        deselect = menu.findItem(R.id.deselect_all);
        applyMenuIconTint(deselect.getIcon());
        deselect.setVisible(false);

        select = menu.findItem(R.id.select_all);
        applyMenuIconTint(select.getIcon());

        if (add.adapter.getItemCount() == 0 || selectedTab == 1)
            select.setVisible(false);

        optionsMenuCreated = true;

        barSearchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                add.filterList(newText);
                return true;
            }
        });


        LifecycleObserver addLifecycleObserver = new LifecycleEventObserver() {
            @Override
            public void onStateChanged(@NonNull LifecycleOwner source, @NonNull Lifecycle.Event event) {
                if (add.adapter != null) {
                    add.getLifecycle().removeObserver(this);

                    add.adapter.selectedItemsCount.observe(CurrenciesSettingsActivity.this, selectedAndTotal -> {
                        toggleElementsVisibility();
                    });
                }

            }
        };

        add.getLifecycle().addObserver(addLifecycleObserver);


        LifecycleObserver editLifecycleObserver = new LifecycleEventObserver() {
            @Override
            public void onStateChanged(@NonNull LifecycleOwner source, @NonNull Lifecycle.Event event) {
                if (edit.adapter != null) {
                    add.getLifecycle().removeObserver(this);

                    edit.adapter.selectedItemsCount.observe(CurrenciesSettingsActivity.this, selectedAndTotal -> {
                        toggleElementsVisibility();
                    });
                }

            }
        };

        edit.getLifecycle().addObserver(editLifecycleObserver);

//        toggleElementsVisibility();

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.deselect_all:
                if (selectedTab == 0)
                    add.adapter.deselectAll();
                else
                    edit.adapter.deselectAll();
                break;
            case R.id.select_all:
                if (selectedTab == 0)
                    add.adapter.selectAll();
                else
                    edit.adapter.selectAll();
                break;
        }
        toggleElementsVisibility();
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        if (changeDone)
            setResult(1);
        else
            setResult(0);
        finish();
        return true;
    }

    public void showDeleteTip() {
        if (!isFirstTimeTooltipShown) {
            return;
        }

        if (deleteTooltip != null) {
            if (!deleteTooltip.isShown()) {
                deleteTooltip.show();
            }
            return;
        }

        deleteTooltip = createStyledSnackbar(
                R.string.tooltip_remove_currency, Snackbar.LENGTH_INDEFINITE);

        deleteTooltip.addCallback(new Snackbar.Callback() {
            @Override
            public void onShown(Snackbar sb) {
                PrefsHelper.setDeleteTooltipShown();
            }

            @Override
            public void onDismissed(Snackbar transientBottomBar, int event) {
                if (event == Snackbar.Callback.DISMISS_EVENT_ACTION
                        || event == Snackbar.Callback.DISMISS_EVENT_SWIPE) {
                    isFirstTimeTooltipShown = false;
                }
            }
        });
        deleteTooltip.setAction(R.string.action_tooltip_got_it, actionView -> {
        });
        deleteTooltip.show();
    }

    private void applyMenuIconTint(Drawable icon) {
        icon.setColorFilter(new PorterDuffColorFilter(
                AndroidHelper.getColorAttribute(this, R.attr.colorAccent),
                PorterDuff.Mode.SRC_ATOP));
    }

    public void toggleElementsVisibility() {
        if (selectedTab == 0) {
            if (add.adapter.isSomethingSelected()) {
                fab.show();
            } else {
                fab.hide();
            }
            select.setVisible(!add.adapter.isAllSelected());
            deselect.setVisible(add.adapter.isSomethingSelected());
        } else {
            boolean selection = edit.adapter.isSelectionMode();
            if (selection) {
                fab.show();
            } else {
                fab.hide();
            }
            select.setVisible(selection && !edit.adapter.isAllSelected());
            deselect.setVisible(selection && edit.adapter.isSomethingSelected());
        }
    }

    public void setNewFabImage(int resId) {
        currentDrawable = resId;
        fab.hide();
    }

    private Snackbar styleSnackbar(Snackbar snackbar) {
        View snackbarView = snackbar.getView();

        MaterialCardView cardView = findViewById(R.id.card_wrapper);
        ViewGroup.MarginLayoutParams cardLayoutParams =
                (ViewGroup.MarginLayoutParams) cardView.getLayoutParams();

        int leftMargin = cardLayoutParams.leftMargin + cardView.getPaddingLeft();
        int rightMargin = cardLayoutParams.rightMargin + cardView.getPaddingRight();
        int height = cardView.getChildAt(0).getMeasuredHeight();

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            height += cardView.getPaddingBottom()
                    + cardView.getStrokeWidth()
                    - (int) cardView.getCardElevation();
            leftMargin -= cardView.getCardElevation() + cardView.getStrokeWidth();
            rightMargin -= cardView.getCardElevation() + cardView.getStrokeWidth();
        } else {
            snackbarView.setElevation(0);
        }

        snackbarView.setMinimumHeight(height);
        snackbarView.setTranslationY(cardView.getPaddingTop() - cardView.getStrokeWidth());

        ViewGroup.MarginLayoutParams snackLayoutParams =
                (ViewGroup.MarginLayoutParams) snackbarView.getLayoutParams();
        snackLayoutParams.leftMargin = leftMargin;
        snackLayoutParams.rightMargin = rightMargin;
        snackbarView.setLayoutParams(snackLayoutParams);

        return snackbar;
    }

    @SuppressWarnings("SameParameterValue")
    private Snackbar createStyledSnackbar(@StringRes int message, int duration) {
        return styleSnackbar(
                SnackbarThemeHelper.createThemedSnackbar(coordinatorLayout, message, duration));
    }

    @SuppressWarnings("SameParameterValue")
    private Snackbar createStyledSnackbar(@NonNull CharSequence message, int duration) {
        return styleSnackbar(
                SnackbarThemeHelper.createThemedSnackbar(coordinatorLayout, message, duration));
    }

    protected void generateUndoSnackbar(List<CurrencyItem> editedItems, boolean added) {

        String message;

        if (editedItems.size() == 1) {
            message = mContext.getString(
                    added
                            ? R.string.message_item_shown
                            : R.string.message_item_hidden,
                    editedItems.get(0).code);
        } else {
            message = mContext.getResources().getQuantityString(
                    added
                            ? R.plurals.message_items_shown
                            : R.plurals.message_items_hidden,
                    editedItems.size(),
                    editedItems.size());
        }

        Snackbar undoSnackbar = createStyledSnackbar(message, Snackbar.LENGTH_LONG);

        undoSnackbar.setAction(R.string.action_undo, view -> {

            Completable
                    .fromRunnable(() -> {
                        Log.d("UndoAction", added ? "hide" : "add");
                        if (added) {
                            model.makeItemsHidden(editedItems);
                        } else {
                            model.makeItemsVisible(editedItems);
                        }
                    }).subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnComplete(() -> {
                        generateUndoSnackbar(editedItems, !added);
                    }).subscribe();
        });

        undoSnackbar.show();
    }

    public void modifyVisualElements(int tabPos) {
        if (tabPos == 0) {
            setNewFabImage(checkDrawable);
            fab.setOnClickListener(v -> {
                ArrayList<CurrencyItem> selectedItems = add.adapter.getSelectedItems();

                Log.d("ToAddCount", String.valueOf(selectedItems.size()));
                Log.d("ToAddCount", String.valueOf(add.adapter.getSelectedCount()));

                changeDone = true;

                Completable
                        .fromRunnable(() ->
                                CurrencyDependencies
                                        .getCurrencyDatabase(mContext)
                                        .currencyDao()
                                        .makeItemsVisible(selectedItems)
                        ).subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .doOnComplete(() ->
                                generateUndoSnackbar(selectedItems, true))
                        .subscribe();

            });
        } else {
            setNewFabImage(deleteDrawable);
            fab.setOnClickListener(v -> {
                List<CurrencyItem> selectedItems = edit.adapter.getSelectedItems();

                changeDone = true;
                edit.adapter.notifyModeChanged(null);

                Completable
                        .fromRunnable(() ->
                                CurrencyDependencies
                                        .getCurrencyDatabase(mContext)
                                        .currencyDao()
                                        .makeItemsHidden(selectedItems)
                        ).subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .doOnComplete(() ->
                                generateUndoSnackbar(selectedItems, false))
                        .subscribe();
            });
        }
    }

    @Override
    public void onBackPressed() {
        if (changeDone)
            setResult(1);
        else
            setResult(0);
        super.onBackPressed();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        tabs.clearOnTabSelectedListeners();
    }

    class CurrencyEditorPager2Adapter extends FragmentStateAdapter {

        String[] categories;
        Fragment[] fragments;

        CurrencyEditorPager2Adapter() {
            super(CurrenciesSettingsActivity.this);
            categories = getResources().getStringArray(R.array.currency_categories);

            List<Fragment> list = CurrenciesSettingsActivity.this.getSupportFragmentManager().getFragments();
            if (list.size() > 0) {
                fragments = new Fragment[list.size()];
                fragments = list.toArray(fragments);
                add = (CurrenciesAddFragment) fragments[0];
                if (list.size() > 1) {
                    edit = (CurrenciesEditFragment) fragments[1];
                }
            }

            if (fragments == null) {
                fragments = new Fragment[]{add, edit};
            }
        }

        @Override
        public int getItemCount() {
            return fragments.length;
        }

        @Nullable
        CharSequence getPageTitle(int position) {
            return categories[position];
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            return fragments[position];
        }
    }

    class CurrencyEditorPagerAdapter extends FragmentPagerAdapter {

        String[] categories;
        Fragment[] fragments;

        CurrencyEditorPagerAdapter(FragmentManager fm) {
            super(fm, FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
            categories = getResources().getStringArray(R.array.currency_categories);

            List<Fragment> list = fm.getFragments();
            if (list.size() > 0) {
                fragments = new Fragment[list.size()];
                fragments = list.toArray(fragments);
                add = (CurrenciesAddFragment) fragments[0];
                edit = (CurrenciesEditFragment) fragments[1];
            }

            if (fragments == null) {
                fragments = new Fragment[]{add, edit};
            }
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            return categories[position];
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            return fragments[position];
        }

        @Override
        public int getCount() {
            return categories.length;
        }
    }

}

File: app/src/main/java/ru/art2000/calculator/view/currency/CurrenciesSettingsActivity.kt
package ru.art2000.calculator.view.currency

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup.MarginLayoutParams
import androidx.activity.viewModels
import androidx.annotation.DrawableRes
import androidx.appcompat.widget.SearchView
import androidx.core.view.updateLayoutParams
import androidx.core.view.updateMargins
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import by.kirich1409.viewbindingdelegate.CreateMethod
import by.kirich1409.viewbindingdelegate.viewBinding
import com.google.android.material.card.MaterialCardView
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.google.android.material.tabs.TabLayoutMediator
import dagger.hilt.android.AndroidEntryPoint
import ru.art2000.calculator.R
import ru.art2000.calculator.databinding.ActivityCurrenciesEditorBinding
import ru.art2000.calculator.model.currency.CurrencyItem
import ru.art2000.calculator.view.AppActivity
import ru.art2000.calculator.view_model.currency.CurrenciesSettingsModel
import ru.art2000.extensions.activities.isLtr
import ru.art2000.extensions.arch.launchAndCollect
import ru.art2000.extensions.arch.launchRepeatOnStarted
import ru.art2000.extensions.collections.LiveList.LiveListObserver
import ru.art2000.extensions.fragments.UniqueReplaceableFragment
import ru.art2000.extensions.views.createThemedSnackbar
import ru.art2000.extensions.writeAndUpdateUi
import ru.art2000.helpers.CurrencyPreferenceHelper
import javax.inject.Inject

@AndroidEntryPoint
class CurrenciesSettingsActivity : AppActivity() {

    private val add by lazy { CurrenciesAddFragment() }
    private val edit by lazy { CurrenciesEditFragment() }

    private var deselect: MenuItem? = null
    private var select: MenuItem? = null

    @DrawableRes
    private val checkDrawable = R.drawable.ic_currencies_done

    @DrawableRes
    private val deleteDrawable = R.drawable.ic_clear_history

    @DrawableRes
    private var currentDrawable = checkDrawable

    private var deleteTooltip: Snackbar? = null

    private val binding by viewBinding<ActivityCurrenciesEditorBinding>(CreateMethod.INFLATE)
    private val model: CurrenciesSettingsModel by viewModels()

    @Inject
    lateinit var prefsHelper: CurrencyPreferenceHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        model.observeAndUpdateDisplayedItems(this)

        launchRepeatOnStarted {
            launchAndCollect(model.removedItems) { generateUndoSnackBar(it, false) }
            launchAndCollect(model.liveIsFirstTimeTooltipShown) { deleteTooltip?.dismiss() }
        }

        val ltr = isLtr

        binding.floatingActionButton.apply {

            addOnShowAnimationListener(object : AnimatorListenerAdapter() {

                override fun onAnimationStart(animator: Animator) {
                    binding.searchViewLayout.updateLayoutParams<MarginLayoutParams> {
                        if (ltr) {
                            rightMargin = binding.coordinator.right -
                                    binding.floatingActionButton.x.toInt()
                        } else {
                            leftMargin = binding.floatingActionButton.right
                        }
                    }
                    setImageResource(currentDrawable)
                }

                override fun onAnimationCancel(animator: Animator) {
                    setImageResource(currentDrawable)
                }
            })

            addOnHideAnimationListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animator: Animator) {
                    if (currentDrawable == deleteDrawable) {
                        showDeleteTip()
                    } else {
                        binding.searchViewLayout.updateLayoutParams<MarginLayoutParams> {
                            if (ltr) {
                                rightMargin = 0
                            } else {
                                leftMargin = 0
                            }
                        }
                    }
                }
            })
        }

        binding.pager2.apply {

            offscreenPageLimit = 2

            registerOnPageChangeCallback(object : OnPageChangeCallback() {
                override fun onPageScrolled(
                    position: Int, positionOffset: Float, positionOffsetPixels: Int
                ) {

                    val maxScroll = measuredWidth
                    val currentScroll = maxScroll * position + positionOffsetPixels
                    binding.searchViewLayout.translationX = currentScroll * if (ltr) -1f else 1f
                    deleteTooltip?.view?.translationX = maxScroll - currentScroll.toFloat()
                }
            })

            val pager2Adapter = CurrencyEditorPager2Adapter()
            adapter = pager2Adapter

            TabLayoutMediator(binding.tabs, this) { tab, position ->
                tab.text = pager2Adapter.getPageTitle(position)
            }.attach()
        }

        binding.tabs.addOnTabSelectedListener(object : OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                when (tab.position) {
                    1 -> {
                        edit.onReplaced(add)
                        add.onReplace(edit)
                    }
                    else -> {
                        add.onReplaced(edit)
                        edit.onReplace(add)
                    }
                }
                model.selectedTab = tab.position
                binding.floatingActionButton.hide()
                modifyVisualElements(tab.position)

                if (lifecycle.currentState.isAtLeast(Lifecycle.State.CREATED)) {
                    toggleElementsVisibility()
                }
                if (tab.position == 1 && !binding.floatingActionButton.isShown) {
                    showDeleteTip()
                } else if (tab.position == 0) {
                    deleteTooltip?.dismiss()
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {}
            override fun onTabReselected(tab: TabLayout.Tab) {
                (if (tab.position == 1) edit else add).onReselected()
            }
        })
    }

    override val clearStatusBar: Boolean
        get() = false

    override val insetAsPadding: Boolean
        get() = false

    override val topViews: List<View>
        get() = listOf(binding.appBar)

    override val bottomViews: List<View>
        get() = listOf(binding.searchViewLayout, binding.floatingActionButton)

    override val leftViews: List<View>
        get() = listOf(binding.root)

    override val rightViews: List<View>
        get() = listOf(binding.root)

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        model.selectedTab = binding.tabs.selectedTabPosition
        modifyVisualElements(binding.tabs.selectedTabPosition)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        binding.toolbar.inflateMenu(R.menu.currencies_editor_menu)

        select = menu.findItem(R.id.select_all)
        deselect = menu.findItem(R.id.deselect_all)

        toggleElementsVisibility()

        binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                model.liveQuery.value = newText
                model.filterHiddenItems(newText)
                return true
            }
        })

        val liveListObserver = object : LiveListObserver<CurrencyItem>() {
            override fun onAnyChanged(previousList: List<CurrencyItem>) {
                toggleElementsVisibility()
            }
        }

        model.selectedHiddenItems.observe(this, liveListObserver)
        model.selectedVisibleItems.observe(this, liveListObserver)
        model.displayedHiddenItems.observe(this, liveListObserver)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.deselect_all -> if (model.selectedTab == 0) {
                model.selectedHiddenItems.clear()
            } else {
                model.isEditSelectionMode = false
                model.selectedVisibleItems.clear()
            }
            R.id.select_all -> if (model.selectedTab == 0) {
                model.completeSelectedHiddenItems()
            } else {
                model.completeSelectedVisibleItems()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    private fun showDeleteTip() {
        if (!model.liveIsFirstTimeTooltipShown.value) return

        deleteTooltip?.also {
            if (!it.isShown) it.show()
            return
        }

        deleteTooltip = binding.coordinator.createThemedSnackbar(
            R.string.tooltip_remove_currency, Snackbar.LENGTH_INDEFINITE
        ).fixedSizes.apply {

            addCallback(object : Snackbar.Callback() {
                override fun onShown(sb: Snackbar) {
                    prefsHelper.setDeleteTooltipShown()
                }

                override fun onDismissed(transientBottomBar: Snackbar, event: Int) {
                    if (event == DISMISS_EVENT_ACTION
                        || event == DISMISS_EVENT_SWIPE
                    ) {
                        model.liveIsFirstTimeTooltipShown.value = false
                    }
                }
            })

            setAction(R.string.action_tooltip_got_it) { }
            show()
        }
    }

    private fun toggleElementsVisibility() {
        if (model.selectedTab == 0) {
            val selectedCount = model.selectedHiddenItems.size
            val displayedCount = model.displayedHiddenItems.size
            if (selectedCount > 0) {
                binding.floatingActionButton.show()
            } else {
                binding.floatingActionButton.hide()
            }
            select?.isVisible = selectedCount < displayedCount
            deselect?.isVisible = selectedCount > 0

            val totalCount = model.hiddenItems.value.size
            binding.searchViewLayout.visibility = if (totalCount > 0) View.VISIBLE else View.GONE
        } else {
            val selectedCount = model.selectedVisibleItems.size
            val totalCount = model.displayedVisibleItems.size

            if (model.isEditSelectionMode && selectedCount > 0) {
                binding.floatingActionButton.show()
            } else {
                binding.floatingActionButton.hide()
            }
            select?.isVisible = model.isEditSelectionMode && selectedCount < totalCount
            deselect?.isVisible = model.isEditSelectionMode && selectedCount > 0
        }
    }

    private fun setNewFabImage(resId: Int) {
        currentDrawable = resId
        binding.floatingActionButton.hide()
    }

    private val Snackbar.fixedSizes: Snackbar
        get() {
            val cardView = findViewById<MaterialCardView>(R.id.card_wrapper)
            val cardLayoutParams = cardView.layoutParams as MarginLayoutParams
            var leftMargin = cardLayoutParams.leftMargin + cardView.paddingLeft
            var rightMargin = cardLayoutParams.rightMargin + cardView.paddingRight
            var bottomMargin = cardLayoutParams.bottomMargin + cardView.paddingBottom
            var height = cardView.getChildAt(0).measuredHeight
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
                height += (cardView.paddingBottom
                        + cardView.strokeWidth
                        - cardView.cardElevation.toInt())
                leftMargin -= (cardView.cardElevation + cardView.strokeWidth).toInt()
                rightMargin -= (cardView.cardElevation + cardView.strokeWidth).toInt()
                bottomMargin -= (cardView.cardElevation + cardView.paddingBottom - cardLayoutParams.topMargin - cardView.strokeWidth).toInt()
            } else {
                view.elevation = 0f
            }
            view.minimumHeight = height

            view.updateLayoutParams<MarginLayoutParams> {
                updateMargins(
                    left = leftMargin,
                    right = rightMargin,
                    bottom = bottomMargin,
                )
            }

            return this
        }

    private fun generateUndoSnackBar(editedItems: List<CurrencyItem>, added: Boolean) {
        if (editedItems.isEmpty()) return

        val message = if (editedItems.size == 1) {
            getString(
                if (added) R.string.message_item_shown else R.string.message_item_hidden,
                editedItems[0].code
            )
        } else {
            resources.getQuantityString(
                if (added) R.plurals.message_items_shown else R.plurals.message_items_hidden,
                editedItems.size,
                editedItems.size
            )
        }

        val undoSnackBar = binding.coordinator
            .createThemedSnackbar(message, Snackbar.LENGTH_LONG).fixedSizes

        undoSnackBar.setAction(R.string.action_undo) {

            writeAndUpdateUi(
                compute = {
                    if (added) {
                        model.makeItemsHidden(editedItems)
                    } else {
                        model.makeItemsVisible(editedItems)
                    }
                },
                update = { generateUndoSnackBar(editedItems, !added) }
            )
        }
        undoSnackBar.show()
    }

    private fun modifyVisualElements(tabPos: Int) {
        if (tabPos == 0) {
            setNewFabImage(checkDrawable)
            binding.floatingActionButton.setOnClickListener {
                val selectedItems = ArrayList(model.selectedHiddenItems)
                writeAndUpdateUi(
                    compute = { model.makeItemsVisible(selectedItems) },
                    update = { generateUndoSnackBar(selectedItems, true) }
                )
            }
        } else {
            setNewFabImage(deleteDrawable)
            binding.floatingActionButton.setOnClickListener {
                val selectedItems = ArrayList(model.selectedVisibleItems)
                writeAndUpdateUi(
                    compute = { model.makeItemsHidden(selectedItems) },
                    update = { generateUndoSnackBar(selectedItems, false) }
                )
            }
        }
    }

    private inner class CurrencyEditorPager2Adapter :
        FragmentStateAdapter(this@CurrenciesSettingsActivity) {

        private val fragments: Array<UniqueReplaceableFragment> = arrayOf(add, edit)

        override fun getItemCount(): Int = fragments.size

        fun getPageTitle(position: Int): CharSequence {
            return fragments[position].getTitle(this@CurrenciesSettingsActivity)
        }

        override fun createFragment(position: Int): Fragment = fragments[position]
    }
}